iwyu_tool_test.py

```diff
@@ -52,6 +52,57 @@ def will_return(self, content):
     def start(self, verbose):
         return MockProcess(self._will_block, self._will_return)
 
+class StubSysExit(object):
+    """ Provides an object that configures an stub behaviour of
+        the sys.exit method. It is possible to modify the value that
+        will be returned
+    """    
+    def __init__(self):
+        self.real_sys_exit = sys.exit
+        sys.exit = lambda x: self._will_return
+        self._will_return = 0
+
+    def reset(self):
+        sys.exit = self.real_sys_exit
+
+    def will_return(self, content):
+        self._will_return = content   
+
+class StubSysArgv(object):
+    """ Provides an object that configures an stub behaviour of
+        the sys.argv method. It is possible to modify the set of
+        values that sys.argv provides.
+    """  
+    def __init__(self, values=[]):
+        self.real_sys_argv = sys.argv
+        sys.argv = values
+
+    def reset(self):
+        sys.argv = self.real_sys_argv
+
+class MockIwyuToolMain(object):
+    """ Provides an object that configures a mock'd behaviour
+        of the iwyu_tool.main method, being possible to obtain
+        the arguments used in the called within the section tested.
+    """     
+    def __init__(self):
+        self.real_iwyu_tool_main = iwyu_tool.main
+        iwyu_tool.main = self._iwyu_tool_main_mock
+        self._will_return = ''
+        self._call_args = []
+
+    def reset(self):
+        iwyu_tool.main = self.real_iwyu_tool_main
+
+    def get_call_args(self):
+        return self._call_args
+
+    def will_return(self, content):
+        self._will_return = content      
+
+    def _iwyu_tool_main_mock(self, *args):
+        self._call_args = args
```

Comment:
This would be stronger if you could remember *which* arg got which value.

I can't see an elegant way to do that, so the easiest is probably to repeat `main`'s argument list here, and store the args in individual named fields;

```
def main(self, compilation_db_path, source_files, verbose, formatter, jobs, iwyu_args):
    self.compilation_db_path = compilation_db_path
    self.source_files = source_files
    self.verbose = verbose
    self.formatter = formatter
    self.jobs = jobs
    self.iwyu_args = iwyu_args
    return 0
```

That way the tests become much more specific, e.g.

    self.assertEqual(['-Xiwyu', '--arg'], mock.iwyu_args);

Reply:
Done ✔️ 
# Details:
I like the idea but i think that finding a mechanism for respecting the (probably changing in time) API may allow others not to fall into waste of time trying to find why tests may fail. 
The solution for this has been to inspect the iwyu_tool.main API and from that point, to create two sepparated attributes for the mock, one containing all these arguments and another one (a dictionary) that will be filled with the arguments of the call. 
Hence, you can get the arguments with something like
```python
       call_args = self.iwyu_tool_main_mock.get_call_args()
       self.assertEqual(['arg1'], call_args['iwyu_args'])
```
I liked the possibility of adding the attributes directly to the mock, but i think that accessing the mock attributes with the name of the arguments considering that they are the arguments used in the last call is less standard.
Let's see how do you see this option but i think that is an elegant approach.